relations/jf/src/note.rs

```diff
@@ -0,0 +1,167 @@
+use ark_ff::Zero;
+use jf_primitives::circuit::rescue::RescueNativeGadget;
+use jf_relation::{Circuit, PlonkCircuit, Variable};
+
+use crate::{
+    shielder_types::{convert_hash, Note, Nullifier, TokenAmount, TokenId, Trapdoor},
+    CircuitField, PlonkResult, PublicInput,
+};
+
+pub enum NoteType {
+    Deposit,
+    Spend,
+}
+
+pub struct SourcedNote {
+    pub note: Note,
+    pub token_id: TokenId,
+    pub token_amount: TokenAmount,
+    pub trapdoor: Trapdoor,
+    pub nullifier: Nullifier,
+    pub note_type: NoteType,
+}
+
+pub struct SourcedNoteVar {
+    pub note_var: Variable,
+    pub token_id_var: Variable,
+    pub token_amount_var: Variable,
+    pub trapdoor_var: Variable,
+    pub nullifier_var: Variable,
+}
+
+pub trait NoteGadget {
+    fn create_note_variable(&mut self, note: &SourcedNote) -> PlonkResult<SourcedNoteVar>;
+    fn enforce_note_preimage(&mut self, note_var: SourcedNoteVar) -> PlonkResult<Variable>;
+}
+
+impl NoteGadget for PlonkCircuit<CircuitField> {
+    fn create_note_variable(&mut self, note: &SourcedNote) -> PlonkResult<SourcedNoteVar> {
+        let note_var = self.create_variable(convert_hash(note.note))?;
+        let token_id_var = self.create_variable(note.token_id.into())?;
+        let token_amount_var = self.create_variable(note.token_amount.into())?;
+        let nullifier_var = self.create_variable(convert_hash(note.nullifier))?;
+        let trapdoor_var = self.create_variable(convert_hash(note.trapdoor))?;
+
+        match note.note_type {
+            NoteType::Spend => {
+                self.set_variable_public(nullifier_var)?;
+            }
+            NoteType::Deposit => {
+                self.set_variable_public(note_var)?;
+                self.set_variable_public(token_id_var)?;
+                self.set_variable_public(token_amount_var)?;
+            }
+        }
+
+        // Ensure that the token amount is valid.
+        // todo: extract token amount limiting to at least constant, or even better to a function/type
+        self.enforce_leq_constant(token_amount_var, CircuitField::from(u128::MAX))?;
+
+        Ok(SourcedNoteVar {
+            note_var,
+            token_id_var,
+            token_amount_var,
+            nullifier_var,
+            trapdoor_var,
+        })
+    }
+
+    fn enforce_note_preimage(&mut self, note_var: SourcedNoteVar) -> PlonkResult<Variable> {
+        let SourcedNoteVar {
+            note_var,
+            token_id_var,
+            token_amount_var,
+            nullifier_var,
+            trapdoor_var,
+        } = note_var;
+
+        let zero_var = self.create_constant_variable(CircuitField::zero())?;
+
+        // Check that the note is valid.
+        let inputs: [usize; 6] = [
+            token_id_var,
+            token_amount_var,
+            trapdoor_var,
+            nullifier_var,
+            zero_var,
+            zero_var,
```

Comment:
Why do we need to add these two `zero_var`s ?

Reply:
jellyfish's rescue hashing function takes 3n arguments... it's messed up and we will provide some wrappers soon

Reply:
b/c similar problems will appear with Merkle trees

Reply:
ok, thanks.